### PR TITLE
Add 24 employment progression post trigger v3

### PR DESCRIPTION
**Pull request by kevin-uk**

### CODE REVIEW

File: NCS.DSS.EmploymentProgression/GetEmploymentProgression/Function/EmploymentProgressionGetTrigger.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            var ApimURL = _httpRequestHelper.GetDssApimUrl(req);
+            if (string.IsNullOrEmpty(ApimURL))
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, "Unable to locate 'apimurl' in request header");
+                return _httpResponseMessageHelper.BadRequest();
+            }
+
+            if (!Guid.TryParse(customerId, out var customerGuid))
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, $"Unable to parse 'customerId' to a Guid: {customerId}");
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/GetEmploymentProgression/Function/EmploymentProgressionGetTrigger.cs]: I have another way to do this check sdd-13 on diversity v3 branch. look at the GuidHelper Class in Common

File: NCS.DSS.EmploymentProgression/GetEmploymentProgressionById/Function/EmploymentProgressionGetByIdTrigger.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            var ApimURL = _httpRequestHelper.GetDssApimUrl(req);
+            if (string.IsNullOrEmpty(ApimURL))
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, "Unable to locate 'apimurl' in request header");
+                return _httpResponseMessageHelper.BadRequest();
+            }
+
+            if (!Guid.TryParse(customerId, out var customerGuid))
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, $"Unable to parse 'customerId' to a Guid: {customerId}");
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/GetEmploymentProgressionById/Function/EmploymentProgressionGetByIdTrigger.cs]: I have another way to do this check sdd-13 on diversity v3 branch. look at the GuidHelper Class in Common

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                                               "<br><b>EmploymentHours:</b> If CurrentEmployment status = 1, 4, 5, 8, 9 then the item must be a valid EmploymentHours reference data item<br>" +
+                                               "<br><b>DateOfEmployment:</b> If CurrentEmployment status = 1, 4, 5, 8, 9 then the item is mandatory, ISO8601:2004 <= datetime.now <br>"
+                                                )]
+        public async Task<HttpResponseMessage> Run([HttpTrigger(AuthorizationLevel.Anonymous, "patch", Route = RouteValue)]HttpRequest req, ILogger logger, string customerId, string EmploymentProgressionId)
+        {
+            _loggerHelper.LogMethodEnter(logger);
+
+            var correlationId = _httpRequestHelper.GetDssCorrelationId(req);
+
+            var guidHelper = new GuidHelper();
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs]: why isnt guid helper injected in?

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+            var ApimURL = _httpRequestHelper.GetDssApimUrl(req);
+            if (string.IsNullOrEmpty(ApimURL))
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, "Unable to locate 'apimurl' in request header");
+                return _httpResponseMessageHelper.BadRequest();
+            }
+
+            if (!Guid.TryParse(customerId, out var customerGuid))
+            {
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs]: could use guid helper here as well

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                return _httpResponseMessageHelper.UnprocessableEntity(JObject.FromObject(new { Error = ex.Message }).ToString());
+            }
+
+            if (employmentProgressionPatchRequest == null)
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, $"A patch body was not provided. CorrelationId: {correlationGuid}.");
+                return _httpResponseMessageHelper.NoContent();
+            }
+
+            _employmentProgressionPatchTriggerService.SetIds(employmentProgressionPatchRequest, employmentProgressionGuid, touchpointId);
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs]: not consistent with the other 14 resources.  Model can set the Id's

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        public async Task<string> GetEmploymentProgressionForCustomerToPatchAsync(Guid customerId, Guid employmentProgressionId)
+        {
+            var employmentProgressionAsString = await _documentDbProvider.GetEmploymentProgressionForCustomerToPatchAsync(customerId, employmentProgressionId);
+
+            return employmentProgressionAsString;
+        }
+
+        public string PatchEmploymentProgressionAsync(string employmentProgressionAsJson, EmploymentProgressionPatch employmentProgressionPatch)
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs]: Not consistent with the other resources. Patch had its own class to handle patching

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            await _serviceBusClient.SendPostMessageAsync(employmentProgression, reqUrl);
+        }
+
+        public bool DoesEmploymentProgressionExistForCustomer(Guid customerId)
+        {
+            return _documentDbProvider.DoesEmploymentProgressionExistForCustomer(customerId);
+        }
+
+        public async Task<bool> DoesCustomerExist(Guid customerId)
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs]: Just reuse the resource Helper no need to duplicate code again.

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            return await _documentDbProvider.DoesCustomerResourceExist(customerId);
+        }
+
+        public void SetIds(EmploymentProgressionPatch employmentProgressionPatch, Guid employmentProgressionGuid, string touchpointId)
+        {
+            employmentProgressionPatch.LastModifiedTouchpointId = touchpointId;
+            employmentProgressionPatch.EmploymentProgressionId = employmentProgressionGuid;
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs]: Move to the model for consistency  set ids and long and lat

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                                              "<br><b>EmploymentHours:</b> If CurrentEmployment status = 1, 4, 5, 8, 9 then the item must be a valid EmploymentHours reference data item<br>" +
+                                              "<br><b>DateOfEmployment:</b> If CurrentEmployment status = 1, 4, 5, 8, 9 then the item is mandatory, ISO8601:2004 <= datetime.now <br>"
+                                                )]
+           public async Task<HttpResponseMessage> Run([HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = RouteValue)]HttpRequest req, ILogger logger, string customerId)
+        {
+            _loggerHelper.LogMethodEnter(logger);
+
+            var correlationId = _httpRequestHelper.GetDssCorrelationId(req);
+
+            var guidHelper = new GuidHelper();
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs]: inject it in

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+            var ApimURL = _httpRequestHelper.GetDssApimUrl(req);
+            if (string.IsNullOrEmpty(ApimURL))
+            {
+                _loggerHelper.LogInformationMessage(logger, correlationGuid, "Unable to locate 'apimurl' in request header");
+                return _httpResponseMessageHelper.BadRequest();
+            }
+
+            if (!Guid.TryParse(customerId, out var customerGuid))
+            {
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs]: use guid helper here

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                _loggerHelper.LogException(logger, correlationGuid, "Unable to retrieve body from req", ex);
+                return _httpResponseMessageHelper.UnprocessableEntity(JObject.FromObject(new { Error = ex.Message }).ToString());
+            }      
+
+            if (employmentProgressionRequest == null)
+            {
+                return _httpResponseMessageHelper.UnprocessableEntity();
+            }
+
+            _employmentProgressionPostTriggerService.SetIds(employmentProgressionRequest, customerGuid, touchpointId);
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs]: move to the model for consistency. Model is the only place which updates the model

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        public async Task<Models.EmploymentProgression> CreateEmploymentProgressionAsync(Models.EmploymentProgression employmentProgression)
+        {
+            if (employmentProgression == null)
+            {
+                return null;
+            }
+
+            employmentProgression.EmploymentProgressionId = Guid.NewGuid();
+
+            if (!employmentProgression.LastModifiedDate.HasValue)
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs]: do this is in SetDefaultValues for consistency once again the model is the only part which updates the model

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        public async Task<Models.EmploymentProgression> CreateEmploymentProgressionAsync(Models.EmploymentProgression employmentProgression)
+        {
+            if (employmentProgression == null)
+            {
+                return null;
+            }
+
+            employmentProgression.EmploymentProgressionId = Guid.NewGuid();
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs]: do this in set ids in the model

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            await _serviceBusClient.SendPostMessageAsync(employmentProgression, reqUrl);
+        }
+
+        public bool DoesEmploymentProgressionExistForCustomer(Guid customerId)
+        {
+            return _documentDbProvider.DoesEmploymentProgressionExistForCustomer(customerId);
+        }
+
+        public void SetIds(Models.EmploymentProgression employmentProgression, Guid customerGuid, string touchpointId)
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs]: Move to Model

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+            employmentProgression.LastModifiedTouchpointId = touchpointId;
+            employmentProgression.CreatedBy = touchpointId;
+
+            if (!employmentProgression.DateProgressionRecorded.HasValue)
+            {
+                employmentProgression.DateProgressionRecorded = DateTime.UtcNow;
+            }
+        }
+
+        public void SetLongitudeAndLatitude(Models.EmploymentProgression employmentProgressionRequest, Position position)
```
Comment (K-Bains) [inline NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs]: Move to Model

File: NCS.DSS.EmploymentProgression.Tests/FunctionTests/EmploymentProgressionsGetTriggerTests.cs [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+using System.Net;
+using NCS.DSS.EmploymentProgression.Tests.FunctionTests.Builders;
+using System.Collections.Generic;
+
+namespace NCS.DSS.EmploymentProgression.Tests.FunctionTests
+{
+    public class EmploymentProgressionsGetTriggerTests
+    {
+        //[Fact]
+        //public async Task Get_WhenTouchPointHeaderIsMission_ReturnBadRequest()
```
[the remark this answers is not in the document]
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression.Tests/FunctionTests/EmploymentProgressionsGetTriggerTests.cs]: Done.

File: NCS.DSS.EmploymentProgression/APIDefinition/GenerateEmploymentProgressionSwaggerDoc.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    public class GenerateEmploymentProgressionSwaggerDoc
+    {
+        public const string ApiTitle = "EmploymentProgressions";
+        public const string ApiDefinitionName = "API-Definition";
+        public const string ApiDefRoute = ApiTitle + "/" + ApiDefinitionName;
+        public const string ApiDescription = "Initial release of Employment Progression";
+        public const string Method = "get";
+
+        private readonly ISwaggerDocumentGenerator _swaggerDocumentGenerator;
+        public const string ApiVersion = "3.0.0";
```
[the remark this answers is not in the document]
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/APIDefinition/GenerateEmploymentProgressionSwaggerDoc.cs]: Done.

File: NCS.DSS.EmploymentProgression/APIDefinition/GenerateEmploymentProgressionSwaggerDoc.cs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+using System.Reflection;
+
+namespace NCS.DSS.EmploymentProgression.APIDefinition
+{
+    public class GenerateEmploymentProgressionSwaggerDoc
+    {
+        public const string ApiTitle = "EmploymentProgressions";
+        public const string ApiDefinitionName = "API-Definition";
+        public const string ApiDefRoute = ApiTitle + "/" + ApiDefinitionName;
+        public const string ApiDescription = "Initial release of Employment Progression";
```
[the remark this answers is not in the document]
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/APIDefinition/GenerateEmploymentProgressionSwaggerDoc.cs]: Done.

File: NCS.DSS.EmploymentProgression/AddEmploymentProgressionServices.cs [hunk 19]
```diff
@@ -0,0 +1,21 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using NCS.DSS.EmploymentProgression.GetEmploymentProgression.Service;
+using NCS.DSS.EmploymentProgression.GetEmploymentProgressionById.Service;
```
[the remark this answers is not in the document]
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/AddEmploymentProgressionServices.cs]: Done.

File: NCS.DSS.EmploymentProgression/Cosmos/Client/CosmosDocumentClient.cs [hunk 20]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+
+namespace NCS.DSS.EmploymentProgression.CosmosDocumentClient
+{
+    public class CosmosDocumentClient : ICosmosDocumentClient
+    {
+        const string AccountEndpointId = "AccountEndpoint=";
```
[the remark this answers is not in the document]
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/Cosmos/Client/CosmosDocumentClient.cs]: This functionality is no in the new CosmosClient NuGet package. By having the this code here this project is protected from a currently incomplete change.

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs (hunk 15 above)
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs]: Agreed to keep functionality in Service.

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs (hunk 11 above)
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/PostEmploymentProgression/Function/EmploymentProgressionPostTrigger.cs]: Agreed to keep functionality in Service.

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs (hunk 5 above)
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Function/EmploymentProgressionPatchTrigger.cs]: Agreed to keep functionality in Service.

File: NCS.DSS.EmploymentProgression/Cosmos/Provider/DocumentDBProvider.cs [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+            }
+            catch (DocumentClientException)
+            {
+                return false;
+            }
+
+            return false;
+        }
+
+        public async Task<bool> DoesCustomerHaveATerminationDate(Guid customerId)
```
[the remark this answers is not in the document]
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/Cosmos/Provider/DocumentDBProvider.cs]: Great will update.

File: NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs (hunk 8 above)
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/PatchEmploymentProgression/Service/EmploymentProgressionPatchTriggerService.cs]: Agreed to keep functionality in Service.

File: NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs (hunk 13 above)
Comment (kevin-uk) [reply NCS.DSS.EmploymentProgression/PostEmploymentProgression/Service/EmploymentProgressionPostTriggerService.cs]: Agreed to keep functionality in Service.

**Pull request merged**